Add Widgets public classes for Core SDK classes: GliaException and RequestCallback

andrews-moc opened the pull request:
**Jira issue:**
[Add public class for com.glia.androidsdk.GliaException to Android Widgets SDK](https://glia.atlassian.net/browse/MOB-4213)

**What was solved?**
- Add public class for `com.glia.androidsdk.GliaException`
- Add public class for `com.glia.androidsdk.RequestCallback`

**Release notes:**
Something general once [MOB-4169](https://glia.atlassian.net/browse/MOB-4169) is done.

 - [x] Feature
 - [ ] Ignore
 - [x] Release notes (Is it clear from the description here?)
 - [ ] Migration guide (If changes are needed for integrator already using the SDK - what needs to be communicated? Add underneath please)

**Additional info:**

 - [x] Is the feature sufficiently tested? All tests fixed? Necessary unit, acceptance, snapshots added? Check that at least new public classes & methods are covered with unit tests
 - [ ] Did you add logging beneficial for troubleshooting of customer issues?
 - [ ] **Did you add new logging?** We would like the logging between platforms to be similar. Refer to [**Logging from Android SDKs** → **Things to consider for newly added logs**](https://glia.atlassian.net/wiki/spaces/ENG/pages/3568861468/Logging+from+Android+SDKs#Things-to-consider-for-newly-added-logs) in Confluence for more information.



[MOB-4169]: https://glia.atlassian.net/browse/MOB-4169?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Andrii-Horishnii-Glia reviewed (commented):
As I see, it means a lot of changes on the integrators' side. Can we just use com.glia.androidsdk.GliaException package and class name in the Widgets SDK?

andrews-moc commented:
> As I see, it means a lot of changes on the integrators' side. Can we just use com.glia.androidsdk.GliaException package and class name in the Widgets SDK?

@Andrii-Horishnii-Glia Thank you for having a look. Could you please explain in more details what exactly do you mean?

DavDo reviewed (changes requested):
I don't think we need to map `GliaException` to `GliaWidgetsException` everywhere.
As far as I understood from the PR, we're mapping them only for internal usage, but in reality we need this mapping only in public interfaces most probably only for `GliaWidgets.onAppCreate()`(if we keep it after @Andrii-Horishnii-Glia's changes, and also for `GliaWidgets.init` function.
For `GliaWidgets.init` function I didn't get it how is the mapping happening, most probably we need to wrap core `Glia.init` call with try/catch and rethrow mapped exception.
I'd suggest doing the same for `RequestCallback`, I'd suggest introducing more client friendly solution for the public functions that require callback like fetching queues and updating visitor info. We're free to introduce new callback and map the core's one inside Widgets.

DavDo commented:
I would suggest making Core internal in Widgets, and then follow the errors in testing app.

andrews-moc commented:
> I would suggest making Core internal in Widgets, and then follow the errors in testing app.

Yes, that's what I'm doing locally. However, it's too early to push this (making Core internal in Widgets) to a branch because a lot of places will be broken and won't compile.

andrews-moc commented:
@DavDo @Andrii-Horishnii-Glia I have taken into account your reasonable feadback and made appropriate changes.
So now I have:
- removed the mapping of `GliaException` to `GliaWidgetsException` in the internal classes
- wrapped the Core SDK calls with try/catch and rethrew mapped `GliaWidgetsException`
- changed the `RequestCallback` interface

Looking forward to seeing your feedback.

DavDo commented on widgetssdk/src/main/java/com/glia/widgets/RequestCallback.kt:
```diff
@@ -0,0 +1,24 @@
+     *
+     * @param response result returned if the request succeeded
+     */
+    fun onSuccess(response: T?)
+
+    /**
+     * Function that is fired if request fails
+     *
+     * @param exception error returned if the request failed
+     */
+    fun onError(exception: GliaWidgetsException?)
```
Why are the `response` and `exception` parameters are nullable?
Do we have cases where the successful response is empty?
If so I'd prefer having separate onSuccess Callback for thouse particular cases.

DavDo replied:
I think we can have separate callbacks for success and error. For example
```kotlin
typealias OnSuccess<T> = (T) -> Unit
typealias OnError = (Throwable?) -> Unit
typealias OnSuccessCallback = () -> Unit
```

andrews-moc replied:
@DavDo Maybe I'm missing something, but `typealias` name can directly be referenced only from Kotlin file, isn't it? We need it to work from Java classes as well. 

So I have created a separate `callbacks` package and put there three interfaces:
1. `OnWidgetsResult<T>`
2. `OnWidgetsSuccess`
3. `OnWidgetsError`

You can see how it looks like for get and update visitor info [here](https://github.com/salemove/android-sdk-widgets/pull/1257/files).

Looking forward to seeing your feedback.

DavDo replied:
You're right!
TypeAliases are not directly accessible from java code, and if we pass it as a function parameter it will be converted to `Function0` or smth similar type, so having separate interfaces will look a way readable.

Andrii-Horishnii-Glia reviewed (approved):
Looks good

DavDo reviewed (approved):
LGTM
Added one suggestion.

gugalo reviewed (commented):
The PR title says "Add public class for GliaException and RequestCallback" however it seems confusing to me. 

Do I understand that instead of `RequestCallback` we now have these three interfaces:
```
OnWidgetsError
OnWidgetsResult
OnWidgetsSuccess
``` 

P.S: I don't have anything against it. Just clarifying that this is intented

andrews-moc commented on widgetssdk/src/main/java/com/glia/widgets/GliaWidgets.java:
```diff
@@ -264,6 +305,11 @@ private static void setupRxErrorHandler() {
     }
 
+    private static RuntimeException mapCoreExceptionToWidgets(GliaException gliaException) {
+        GliaWidgetsException widgetsException = GliaWidgetsException.Companion.from(gliaException);
```
Yes, indeed. Thank you. Fixed. 68067f226a6e4e6544df17402f4f42af47e2836c

andrews-moc commented:
> The PR title says "Add public class for GliaException and RequestCallback" however it seems confusing to me.
> 
> Do I understand that instead of `RequestCallback` we now have these three interfaces:
> 
> ```
> OnWidgetsError
> OnWidgetsResult
> OnWidgetsSuccess
> ```
> 
> P.S: I don't have anything against it. Just clarifying that this is intented

Yes, that's correct.
To avoid the confusion, I've just renamed PR to 'Add Widgets public classes for Core SDK classes: GliaException and RequestCallback'.

andrews-moc commented on widgetssdk/src/main/java/com/glia/widgets/GliaWidgetsException.kt:
```diff
@@ -0,0 +1,52 @@
+class GliaWidgetsException(
+    val debugMessage: String?,
+    val gliaCause: Cause?
+) : RuntimeException(
```
Done a6ee8fe588448cf46ad28a26957e357155732974

andrews-moc commented on widgetssdk/src/main/java/com/glia/widgets/GliaWidgetsException.kt:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * Possible exception causes
+     */
+    enum class Cause(val sdkCause: GliaCoreException.Cause?) {
```
@gugalo I agree with this point.
Changed, please have a look: 62017229c64c966917f0cebfc1fdc4f1e5044404
I've also added unit tests.
